compiler/src/dotty/tools/dotc/typer/QuotesAndSplices.scala

```diff
@@ -325,7 +319,51 @@ trait QuotesAndSplices {
         new TreeTypeMap(typeMap = typeMap).transform(shape1)
       }
 
-    (typeBindings, shape2, patterns)
+    (bindSymMapping, shape2, patterns)
+  }
+
+  /** For each type variable defined in the quote pattern we generate an equivalent
+   *  binding that will be as type variable in the encoded `unapply` of the quote pattern.
+   *
+   *  @return Mapping from type variable symbols defined in the quote pattern into
+   *          type variable `Bind` definitions for the `unapply` of the quote pattern.
+   *          This mapping retains the original type variable definition order.
+   */
+  private def unapplyBindingsMapping(quoted: Tree)(using Context): collection.Map[Symbol, Bind] = {
+    val mapping = mutable.LinkedHashMap.empty[Symbol, Symbol]
+    // Collect all existing type variable bindings and create new symbols for them.
+    // The old info is used, it may contain references to the old symbols.
+    new tpd.TreeTraverser {
+      def traverse(tree: Tree)(using Context): Unit = tree match {
+        case _: SplicePattern =>
+        case Select(pat: Bind, _) if tree.symbol.isTypeSplice =>
+          val sym = tree.tpe.dealias.typeSymbol
+          if sym.exists then registerNewBindSym(sym)
+        case tdef: TypeDef  =>
+          if tdef.symbol.hasAnnotation(defn.QuotedRuntimePatterns_patternTypeAnnot) then
+            registerNewBindSym(tdef.symbol)
+          traverseChildren(tdef)
+        case _ =>
+          traverseChildren(tree)
+      }
+      private def registerNewBindSym(sym: Symbol): Unit =
+        if !mapping.contains(sym) then
+          mapping(sym) = newSymbol(ctx.owner, sym.name.toString.stripPrefix("$").toTypeName, Case | sym.flags, sym.info, coord = quoted.span)
+    }.traverse(quoted)
+
+    // Replace symbols in `mapping` in the infos of the new symbol and register GADT bounds.
+    // GADT bounds need to be added after the info is updated to avoid references to the old symbols.
+    var oldBindings: List[Symbol] = mapping.keys.toList
+    var newBindingsRefs: List[Type] = mapping.values.toList.map(_.typeRef)
```

Comment:
Look like those don't need to be vars?
```suggestion
    val oldBindings: List[Symbol] = mapping.keys.toList
    val newBindingsRefs: List[Type] = mapping.values.toList.map(_.typeRef)
```